### ISSUE TITLE
Incorrect distances returned for all-zero query

### ISSUE THREAD

**Issue by lsorber** [nmslib/nmslib#321]
An all-zero query vector will result in NMSLib incorrectly reporting a distance of zero for its nearest neighbours (see example below). Is this related to #187? Is there a suggested workaround?

```python
# Training set (CSR sparse matrix)
X.todense()
# Out:
# matrix([[4., 2., 3., 1., 0., 0., 0., 0., 0.],
#         [2., 1., 0., 0., 3., 0., 1., 2., 1.],
#         [4., 2., 0., 0., 3., 1., 0., 0., 0.]], dtype=float32)

# Query vector (CSR sparse matrix)
r.todense()
# Out:
# matrix([[0., 0., 0., 0., 0., 0., 0., 0., 0.]], dtype=float32)

# Train and query
import nmslib
index = nmslib.init(
    method='hnsw',
    space='cosinesimil_sparse_fast',
    data_type=nmslib.DataType.SPARSE_VECTOR,
    dtype=nmslib.DistType.FLOAT)
index.addDataPointBatch(X)
index.createIndex()
index.knnQueryBatch(r, k=3)
# Out:
# [(array([2, 1, 0], dtype=int32), array([0., 0., 0.], dtype=float32))]

# Note that distances are all 0, which is incorrect!
# Same result for dense training & query vectors.
```

**Comment by searchivarius** [comment]
is it the latest NMSLIB version?

**Comment by lsorber** [comment]
Yes, 1.7.2

**Comment by searchivarius** [comment]
@lsorber I cannot reproduce it without a complete script/data set. Could you create one? Thanks!

**Comment by lsorber** [comment]
Here's a minimal example that reproduces the problem:
```python
import nmslib
import numpy as np

X = np.array([
    [4., 2., 3., 1., 0., 0., 0., 0., 0.],
    [2., 1., 0., 0., 3., 0., 1., 2., 1.],
    [4., 2., 0., 0., 3., 1., 0., 0., 0.]], dtype=np.float32)
r = np.array([
    [0., 0., 0., 0., 0., 0., 0., 0., 0.]], dtype=np.float32)

index = nmslib.init(method='hnsw', space='cosinesimil')
index.addDataPointBatch(X)
index.createIndex()
index.knnQueryBatch(r, k=3)
# Prints: [(array([1, 2, 0], dtype=int32), array([0., 0., 0.], dtype=float32))]
```
The distances are incorrectly returned as 0. Same results for sparse matrices `X` and `r` too.

**Comment by lsorber** [comment]
I think it should be NaN (unless perhaps there is also an all-zero vector in the training set). In our application, vectors have integer valued entries, and under those circumstances it is not uncommon to see all-zero query vectors.

**Comment by searchivarius** [comment]
@lsorber sorry, I don't think so. One shouldn't compute cosine for all-zero vectors, it's not defined.

**Comment by searchivarius** [comment]
PS: if you produce NaNs it gets ugly one way or another later.

**Comment by lsorber** [comment]
> One shouldn't compute cosine for all-zero vectors, it's not defined.
I understand your point of view, but if you do happen to compute the cosine distance for an all-zero vector, Nmslib will currently (1) not inform the user of any issue and (2) silently return an incorrect distance.

I guess it comes down to who's responsible for validating user input and your point of view is it's the user.

> PS: if you produce NaNs it gets ugly one way or another later.
Actually, in our case it makes a lot of sense: if you filter the nearest neighbours on a max distance (e.g., dist < max_dist), that would fail for a NaN dist and so wouldn't return any candidates. With an incorrect dist of zero it looks like there are as many "perfect" nearest neighbour matches as there are vectors in the training set.

**Comment by yurymalkov** [comment]
@lsorber Is there any knn libraries with a different point of view for this issue?

**Comment by lsorber** [comment]
@yurymalkov Sklearn returns a distance of 1 for an all-zero query vector (see example below). This is because they choose not to normalize the query vector if it's norm is zero [1]. As a result, the inner product between the training vectors and query vector is zero, and the distance returned is 1. I think NaN is the correct result, but would still prefer receiving an "orthogonal" distance of 1 over a "perfect match" distance of 0.

```python
from sklearn.neighbors import NearestNeighbors
import numpy as np

X = np.array([
    [4., 2., 3., 1., 0., 0., 0., 0., 0.],
    [2., 1., 0., 0., 3., 0., 1., 2., 1.],
    [4., 2., 0., 0., 3., 1., 0., 0., 0.]], dtype=np.float32)
r = np.array([
    [0., 0., 0., 0., 0., 0., 0., 0., 0.]], dtype=np.float32)

nn = NearestNeighbors(n_neighbors=3, algorithm='brute', metric='cosine')
nn.fit(X)
distances, indices = nn.kneighbors(r)
distances
# Prints: array([[1., 1., 1.]], dtype=float32)
```

[1] https://github.com/scikit-learn/scikit-learn/blob/0.19.1/sklearn/preprocessing/data.py#L61

**Comment by searchivarius** [comment]
@lsorber perhaps, the best way to deal with it is to throw an exception, unless this has some serious performance implications (though I doubt). However, we can also ensure that all-zero vectors are never nearest neighbors of anything except maybe when you have to all-zero vectors. So, is this what you suggest?

**Comment by lsorber** [comment]
@searchivarius It's your call, as discussed there are many options:
- Raise an exception
- Return a NaN distance
- Return a distance of 1 like Sklearn does

I'm still voting for NaN, but would be happy with any solution.

**Comment by searchivarius** [comment]
@lsorber it may not be even possible to detect NaNs: 
http://searchivarius.org/blog/gcc-disables-isnan-and-isinf-when-compiling-ffast-math-flag

**Comment by lsorber** [comment]
That doesn't sound like a good situation indeed, but it looks like it may be possible to check for NaNs even when using `-ffast-math` based on [this SO question](https://stackoverflow.com/questions/38278300/successfully-enabling-fno-finite-math-only-on-nan-removal-method).

**Comment by searchivarius** [comment]
@lsorber I now just use my own functions to check for NaNs, but anyways I don't that purposeful generation of NANs INFs denormalized numbers and other wicked magic animals is a good thing to do.

**Comment by searchivarius** [comment]
PS: We should make the phrase no NaNsensical results as our motto :-)

**Comment by lsorber** [comment]
After giving this some more thought, the "exception" solution is probably not a good choice. For example, if I batch query several valid vectors and one all-zero vector, then the whole batch would fail because of the single all-zero vector.

**Comment by searchivarius** [comment]
@lsorber so, what do we really need: return the largest possible distance value? what do we do when we compare two all-zero vectors?

**Comment by lsorber** [comment]
Maybe the easiest solution is the sklearn solution of not normalising the vectors if their norm is zero. The output will be compatible with sklearn that way, which is a plus if you want to wrap nmslib as an sklearn Estimator.

**Comment by searchivarius** [comment]
@lsorber nmslib doesn't really normalize vectors: the question is what distance should be if you have
d(0, x) where x != 0 and d(0, 0)
We can make d(0,x) = 2 for the cosine distance, which is reasonable, by what should we use for d(0, 0)?

**Comment by lsorber** [comment]
I would prefer to see d(0,x) = 1 (regardless of whether there is normalization) to be compatible with sklearn. The interpretation to me is that the all-zero vector is orthogonal to x, which _can_ be true for many x. For d(0,x) = 2, the interpretation would be that 0 is the opposite of x, which _cannot_ be true for many x.

Then there's the choice of what d(0,0) should be. A reasonable option would be 0, but I can see 1 being reasonable as well (similar to how NaN != NaN). To break the tie, I would again go with what sklearn does, which is return 1.

**Comment by searchivarius** [comment]
This is fixed and there is a unit test to prevent a regression. All-zero query vector produces the distance 1.
